File: sdk/ruby/lib/bsdkrun/volumes.rb
# frozen_string_literal: true

require "json"

module Bsdkrun
  # Host-level volume operations.
  module Volumes
    module_function

    # List persistent volumes.
    # @return [Array<VolumeInfo>]
    def list
      res = Process.run!(["volume", "ls", "--json"], label: "bsdkrun volume ls")
      rows = JSON.parse(res.stdout.empty? ? "[]" : res.stdout)
      rows.map { |row| VolumeInfo.from_row(row) }
    end

    # Remove one or more volumes (and their data).
    #
    # @param names [String, Array<String>]
    # @param force [Boolean]
    # @return [void]
    def remove(names, force: false)
      args = ["volume", "rm"]
      args << "--force" if force
      args.concat(Array(names))
      Process.run!(args, label: "bsdkrun volume rm")
      nil
    end
  end
end

File: sdk/clojure/src/bsdkrun/volumes.clj
(ns bsdkrun.volumes
  "Host-level volume operations. Mirrors `sdk/ruby/lib/bsdkrun/volumes.rb`."
  (:refer-clojure :exclude [list])
  (:require [bsdkrun.process :as process]
            [bsdkrun.types :as types]
            [bsdkrun.util :as util]))

(defn list
  "List persistent volumes."
  []
  (let [res (process/run! ["volume" "ls" "--json"] {:label "bsdkrun volume ls"})]
    (mapv types/volume-info-from-row (types/read-json-rows (:stdout res)))))

(defn remove!
  "Remove one or more volumes (and their data). `names` is a string or a
  vector of strings. `{:force true}` skips confirmation."
  ([names] (remove! names {}))
  ([names {:keys [force]}]
   (let [argv (cond-> ["volume" "rm"] force (conj "--force"))
         argv (into argv (util/as-seq names))]
     (process/run! argv {:label "bsdkrun volume rm"})
     nil)))
